### ISSUE TITLE
Make `tsc` build node "module" compliant code

### ISSUE THREAD

**Issue by homersimpsons** [microsoft/TypeScript#56781]
### 🔍 Search Terms

Reproduction: https://github.com/homersimpsons/typescript-node20-module

TL;DR: TypeScript cannot build Node20 + `"type":"module"` compliant code

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

With `compilerOptions.target = "ESNext"`, TypeScript should output node-compliant code

### 📃 Motivating Example

Improve Developer Experience when using TypeScript with `node` and `ECMAScript module` mode

### 💻 Use Cases

1. What do you want to use this for?
  To easily compile TypeScript to real JavaScript code
2. What shortcomings exist with current approaches?
  They all involve either another tool (with possibly runtime overhead), or weird import declarations
3. What workarounds are you using in the meantime?
  For now, I reverted to using Node18 with the `--es-module-specifier-resolution=node`.

**Comment by fatcerberus** [comment]
You haven’t described, concretely, what the issue is.  How does module resolution etc. differ in Node 20 compared to previous versions, and why is it a problem?

One thing I notice right off the bat is that the `tsconfig.json` in the linked repro doesn't specify [`node16`/`nodenext` module resolution](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext).

**Comment by fatcerberus** [comment]
> In fact, TypeScript already does the job to resolve the path, but it just does not output it.

This just means the project is misconfigured (see above note on node16/nodenext options).

TS never rewrites import paths.  There are many issues asking for this behavior and they’ve all been rejected by the maintainers.  If you want to use ESM in Node then you have to include the extension in the import specifier, period.  That’s not a bug.

**Comment by nmain** [comment]
When the project settings are fixed to reflect the actual runtime being executed against, `tsc` understands and flags the error:

> Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean './utils.js'? ts(2835)

So this seems to be just a duplicate of https://github.com/microsoft/TypeScript/issues/16577.

**Comment by homersimpsons** [comment]
> You haven’t described, concretely, what the issue is. How does module resolution etc. differ in Node 20 compared to previous versions, and why is it a problem?

Sorry I was not clear, the issue is that `import ... from '...';` are not working after compilation to JavaScript.

The imports have to be rewritten:
```diff
- import { add } from './utils'
+ import { add } from './utils.js'
```

> This just means the project is misconfigured (see above note on node16/nodenext options).

Adding `"module": "NodeNext"` forces me to add the `.js` extension to my `import` statements.

I find this wrong as there is no `.js` file in the source. I would be okay with a `.ts` extension in my import statements.

> TS never rewrites import paths. There are many issues asking for this behavior and they’ve all been rejected by the maintainers. If you want to use ESM in Node then you have to include the .js extension in the import specifier, period. That’s not a bug.

I find it frustrating that TypeScript do not output valid JavaScript code if the developer does not use "unresolved" imports.

Furthermore, it does not allow `import ... from '*.ts'` with path rewrite during the compilation.

> So this seems to be just a duplicate of https://github.com/microsoft/TypeScript/issues/16577.

Exactly, I could not seem to find it, thank you. I will read it carefully.

## Conclusion

So I guess I'm left with writing `*.js` to my import path or using a tool to write them for me ...

**Comment by fatcerberus** [comment]
> I find it frustrating that TypeScript do not output valid JavaScript code if the developer does not use "unresolved" imports.

The way to think about this is that it doesn't work because you didn't *write* valid JavaScript code in the first place; TS simply left what you wrote as-is.  TypeScript as a programming language doesn't really exist in a vacuum, it's more like... an expansion pack for JS that gives you static type checking.  So whatever import you would have written if you were writing pure JS, you also have to write it the same way in TS.  That's just the way it is.

**Comment by RyanCavanaugh** [comment]
This is not an oversight. See https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-specifiers-are-not-transformed

**Comment by homersimpsons** [comment]
>> I find it frustrating that TypeScript do not output valid JavaScript code if the developer does not use "unresolved" imports.

> The way to think about this is that it doesn't work because you didn't write valid JavaScript code in the first place; TS simply left what you wrote as-is. TypeScript as a programming language doesn't really exist in a vacuum, it's more like... an expansion pack for JS that gives you static type checking. So whatever import you would have written if you were writing pure JS, you also have to write it the same way in TS. That's just the way it is.

I agree with this philosophy and this is why I closed this issue. But I find this confusing to have to write `import ... from '....js'` while I have 0 js file in my project.

I understand the philosophy that "TypeScript" adapts to your code and your code is JS compatible.
